File: Bookcase core/src/test/java/cz/vhromada/book/facade/impl/CategoryFacadeImplTest.java
package cz.vhromada.book.facade.impl;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import cz.vhromada.book.entity.Category;
import cz.vhromada.book.utils.CategoryUtils;
import cz.vhromada.common.facade.MovableParentFacade;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.facade.MovableParentFacadeTest;
import cz.vhromada.common.validator.MovableValidator;
import cz.vhromada.converter.Converter;

import org.junit.jupiter.api.Test;

/**
 * A class represents test for class {@link CategoryFacadeImpl}.
 *
 * @author Vladimir Hromada
 */
class CategoryFacadeImplTest extends MovableParentFacadeTest<Category, cz.vhromada.book.domain.Category> {

    /**
     * Test method for {@link CategoryFacadeImpl#CategoryFacadeImpl(MovableService, Converter, MovableValidator)} with null service for categories.
     */
    @Test
    void constructor_NullCategoryService() {
        assertThatThrownBy(() -> new CategoryFacadeImpl(null, getConverter(), getMovableValidator())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link CategoryFacadeImpl#CategoryFacadeImpl(MovableService, Converter, MovableValidator)} with null converter.
     */
    @Test
    void constructor_NullConverter() {
        assertThatThrownBy(() -> new CategoryFacadeImpl(getMovableService(), null, getMovableValidator())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link CategoryFacadeImpl#CategoryFacadeImpl(MovableService, Converter, MovableValidator)} with null validator for category.
     */
    @Test
    void constructor_NullCategoryValidator() {
        assertThatThrownBy(() -> new CategoryFacadeImpl(getMovableService(), getConverter(), null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableParentFacade<Category> getMovableParentFacade() {
        return new CategoryFacadeImpl(getMovableService(), getConverter(), getMovableValidator());
    }

    @Override
    protected Category newEntity(final Integer id) {
        return CategoryUtils.newCategory(id);
    }

    @Override
    protected cz.vhromada.book.domain.Category newDomain(final Integer id) {
        return CategoryUtils.newCategoryDomain(id);
    }

    @Override
    protected Class<Category> getEntityClass() {
        return Category.class;
    }

    @Override
    protected Class<cz.vhromada.book.domain.Category> getDomainClass() {
        return cz.vhromada.book.domain.Category.class;
    }

}

File: bookcase-core/src/test/kotlin/cz/vhromada/bookcase/facade/impl/CategoryFacadeImplTest.kt
package cz.vhromada.bookcase.facade.impl

import com.nhaarman.mockitokotlin2.any
import cz.vhromada.bookcase.entity.Category
import cz.vhromada.bookcase.utils.CategoryUtils
import cz.vhromada.common.facade.MovableParentFacade
import cz.vhromada.common.test.facade.MovableParentFacadeTest

/**
 * A class represents test for class [CategoryFacadeImpl].
 *
 * @author Vladimir Hromada
 */
class CategoryFacadeImplTest : MovableParentFacadeTest<Category, cz.vhromada.bookcase.domain.Category>() {

    override fun getFacade(): MovableParentFacade<Category> {
        return CategoryFacadeImpl(service, mapper, validator)
    }

    override fun newEntity(id: Int?): Category {
        return CategoryUtils.newCategory(id)
    }

    override fun newDomain(id: Int?): cz.vhromada.bookcase.domain.Category {
        return CategoryUtils.newCategoryDomain(id)
    }

    override fun anyEntity(): Category {
        return any()
    }

    override fun anyDomain(): cz.vhromada.bookcase.domain.Category {
        return any()
    }

}
